In [ ]:
import pandas as pd
import numpy as np
import torch
from transformers import (
    AutoModelForCausalLM,
    AutoTokenizer,
    TrainingArguments,
    Trainer,
    DataCollatorForLanguageModeling
)
from datasets import Dataset
from sklearn.metrics import accuracy_score, precision_recall_fscore_support, roc_auc_score
from sklearn.model_selection import train_test_split
from peft import LoraConfig, get_peft_model, prepare_model_for_kbit_training
import time
import json
import random
from tqdm.auto import tqdm


In [ ]:
# Доли пропущенных признаков для экспериментов
MISSING_RATES = [0.0, 0.2, 0.5, 0.9]

# Настройка промпта для Bank
prompt_config = {
        "task": "Predict whether a bank client will subscribe",
        "pos_label": "yes",
        "neg_label": "no",
        "entity": "subscription",
        "question": "Will this client subscribe?"
}

FEATURE_MAPPINGS = {
        'V1': 'Age',
        'V2': 'Job',
        'V3': 'Martial',
        'V4': 'Education',
        'V5': 'Default',
        'V6': 'Balance',
        'V7': 'Housing',
        'V8': 'Loan',
        'V9': 'Contact',
        'V10': 'Day of Week',
        'V11': 'Month',
        'V12': 'Duration',
        'V13': 'Campaign',
        'V14': 'Pdays',
        'V15': 'Previous',
        'V16': 'Poutcome',
        'Class': 'y'
}

openml_id = 1461

base_output_dir = "/content/drive/MyDrive/finetuned_qwen_missing_Bank"

# 1. Загрузка данных

In [ ]:
import pandas as pd
from sklearn.datasets import fetch_openml
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

def load_dataset(openml_id=1464):
    dataset = fetch_openml(data_id=openml_id, as_frame=True, parser='auto')
    X = dataset.data
    y = dataset.target

    df = X.copy()
    feature_names = X.columns.to_list()

    df = df.rename(columns=FEATURE_MAPPINGS)
    feature_names = list(FEATURE_MAPPINGS.values())[:-1]
    target_name = list(FEATURE_MAPPINGS.values())[-1]
    df[target_name] = y

    # Преобразование целевой переменной в бинарный формат
    if y.dtype == 'object' or y.dtype.name == 'category':
        le = LabelEncoder()
        df[target_name] = le.fit_transform(df[target_name])

    return df, feature_names, target_name


def split_dataset(df, target_name, test_size=0.2, val_size=0.25, seed=42):

    # Разделение на train/val/test (60/20/20)
    train_val_df, test_df = train_test_split(
        df,
        test_size=test_size,
        random_state=seed,
        stratify=df[target_name]
    )

    train_df, val_df = train_test_split(
        train_val_df,
        test_size=val_size,
        random_state=seed,
        stratify=train_val_df[target_name]
    )

    return train_df, val_df, test_df

df, feature_names, target_name = load_dataset(openml_id)
train_df, val_df, test_df = split_dataset(df, target_name)

df.head(5)

,Age,Job,Martial,Education,Default,Balance,Housing,Loan,Contact,Day of Week,Month,Duration,Campaign,Pdays,Previous,Poutcome,y
0,58,management,married,tertiary,no,2143,yes,no,unknown,5,may,261,1,-1,0,unknown,0
1,44,technician,single,secondary,no,29,yes,no,unknown,5,may,151,1,-1,0,unknown,0
2,33,entrepreneur,married,secondary,no,2,yes,yes,unknown,5,may,76,1,-1,0,unknown,0
3,47,blue-collar,married,unknown,no,1506,yes,no,unknown,5,may,92,1,-1,0,unknown,0
4,33,unknown,single,unknown,no,1,no,no,unknown,5,may,198,1,-1,0,unknown,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45211 entries, 0 to 45210
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   Age          45211 non-null  int64   
 1   Job          45211 non-null  category
 2   Martial      45211 non-null  category
 3   Education    45211 non-null  category
 4   Default      45211 non-null  category
 5   Balance      45211 non-null  int64   
 6   Housing      45211 non-null  category
 7   Loan         45211 non-null  category
 8   Contact      45211 non-null  category
 9   Day of Week  45211 non-null  int64   
 10  Month        45211 non-null  category
 11  Duration     45211 non-null  int64   
 12  Campaign     45211 non-null  int64   
 13  Pdays        45211 non-null  int64   
 14  Previous     45211 non-null  int64   
 15  Poutcome     45211 non-null  category
 16  y            45211 non-null  int64   
dtypes: category(9), int64(8)
memory usage: 3.1 MB


In [ ]:
for name, df in [("train", train_df), ("val", val_df), ("test", test_df)]:
    counts = df[target_name].value_counts()
    pcts   = df[target_name].value_counts(normalize=True) * 100
    print(f"\n{name} (всего: {len(df)}):")
    print(f"  {prompt_config['neg_label']}: {counts[0]:5d} — {pcts[0]:.1f}%")
    print(f"  {prompt_config['pos_label']}: {counts[1]:5d} — {pcts[1]:.1f}%")


train (всего: 27126):
  no: 23953 — 88.3%
  yes:  3173 — 11.7%

val (всего: 9042):
  no:  7984 — 88.3%
  yes:  1058 — 11.7%

test (всего: 9043):
  no:  7985 — 88.3%
  yes:  1058 — 11.7%


# 2. Вспомогательные функции

In [ ]:
def row_to_text_template_with_missing(row, feature_names, target_name, missing_rate=0.0, seed=None):

    rng = np.random.default_rng(seed)

    # Определяем, какие признаки пропустить
    n_drop = int(len(feature_names) * missing_rate)
    dropped = set(rng.choice(feature_names, size=n_drop, replace=False)) if n_drop > 0 else set()

    template_parts = []

    for feature in feature_names:
        if feature in dropped:
            continue  # пропускаем признак

        value = row[feature]

        if isinstance(value, (int, np.integer)):
            phrase = f"The value of {feature} is {value}."
        elif isinstance(value, (float, np.floating)):
            phrase = f"The value of {feature} is {value:.2f}."
        else:
            phrase = f"The category of {feature} is {value}."

        template_parts.append(phrase)

    text = " ".join(template_parts)

    return text

# Тест
test_row = train_df.iloc[0]
display(train_df.head(1))
for rate in MISSING_RATES:
    text_output = row_to_text_template_with_missing(
        row=test_row,
        feature_names=feature_names,
        target_name=target_name,
        missing_rate=rate
    )

    print(f"Missing Rate: {rate * 100:.0f}%")
    print(text_output)

,Age,Job,Martial,Education,Default,Balance,Housing,Loan,Contact,Day of Week,Month,Duration,Campaign,Pdays,Previous,Poutcome,y
10333,43,services,married,secondary,no,2478,yes,no,unknown,12,jun,157,1,-1,0,unknown,0


Missing Rate: 0%
The value of Age is 43. The category of Job is services. The category of Martial is married. The category of Education is secondary. The category of Default is no. The value of Balance is 2478. The category of Housing is yes. The category of Loan is no. The category of Contact is unknown. The value of Day of Week is 12. The category of Month is jun. The value of Duration is 157. The value of Campaign is 1. The value of Pdays is -1. The value of Previous is 0. The category of Poutcome is unknown.
Missing Rate: 20%
The value of Age is 43. The category of Martial is married. The category of Education is secondary. The category of Default is no. The value of Balance is 2478. The category of Housing is yes. The category of Loan is no. The value of Day of Week is 12. The category of Month is jun. The value of Campaign is 1. The value of Pdays is -1. The value of Previous is 0. The category of Poutcome is unknown.
Missing Rate: 50%
The value of Age is 43. The category of Job 

In [ ]:
def parse_prediction(response, prompt_config):
    response = response.lower().strip().rstrip('.,!? ')
    pos, neg = prompt_config['pos_label'].lower(), prompt_config['neg_label'].lower()

    # Точное совпадение
    if response == pos: return 1
    if response == neg: return 0

    # Начинается с метки
    if response.startswith(pos): return 1
    if response.startswith(neg): return 0

    # Содержит метку как отдельное слово
    words = response.split()
    if pos in words: return 1
    if neg in words: return 0

    # Не удалось распознать
    print(f"Warning: Could not parse '{response}' (expected '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}')")
    return 0



response = prompt_config['pos_label']
pred = parse_prediction(response, prompt_config)
print(f"Response: '{response}'\nPrediction: {pred}\n")

response = "hi"
pred = parse_prediction(response, prompt_config)
print(f"Response: '{response}'\nPrediction: {pred}")

Response: 'yes'
Prediction: 1

Response: 'hi'
Prediction: 0


In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
# Вычисление метрик качества
def compute_metrics(y_true, y_pred, y_prob=None):
    roc = roc_auc_score(y_true, y_prob if y_prob is not None else y_pred)
    return {
        "ROC-AUC": roc_auc_score(y_true, y_prob if y_prob is not None else y_pred),
        "F1":       f1_score(y_true, y_pred, zero_division=0),
        "Accuracy": accuracy_score(y_true, y_pred),
        "Precision": precision_score(y_true, y_pred, zero_division=0),
        "Recall":   recall_score(y_true, y_pred, zero_division=0),
    }

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.utils import resample

def bootstrap_metrics(y_true, y_pred, y_prob=None, n_iter=1000):
    scores = []

    for i in range(n_iter):
        # Bootstrap выборка
        if y_prob is not None:
            y_true_boot, y_pred_boot, y_prob_boot = resample(
                y_true, y_pred, y_prob, random_state=i+1
            )
        else:
            y_true_boot, y_pred_boot = resample(
                y_true, y_pred, random_state=i+1
            )
            y_prob_boot = None

        try:
            # Вычисление метрик
            if y_prob_boot is not None:
                auc = roc_auc_score(y_true_boot, y_prob_boot)
            else:
                auc = 0.0

            f1 = f1_score(y_true_boot, y_pred_boot, zero_division=0)
            pr = precision_score(y_true_boot, y_pred_boot, zero_division=0)
            rc = recall_score(y_true_boot, y_pred_boot, zero_division=0)

            acc = accuracy_score(y_true_boot, y_pred_boot)
            scores.append((auc, f1, acc, pr, rc))

        except ValueError:
            # Пропускание сэмплов, где не представлены все классы
            continue

    scores = np.asarray(scores)
    means, stds = scores.mean(0), scores.std(0, ddof=1)
    names = ["ROC-AUC", "F1", "Accuracy", "Precision", "Recall"]

    return {n: f"{m:.4f}±{s:.4f}" for n, m, s in zip(names, means, stds)}

# Тестирование функции
sample_row = test_df.iloc[0]
sample_text = row_to_text_template_with_missing(sample_row, feature_names, target_name)
print(f"\nПример текста:\n{sample_text[:300]}")


Пример текста:
The value of Age is 40. The category of Job is blue-collar. The category of Martial is married. The category of Education is primary. The category of Default is no. The value of Balance is 640. The category of Housing is yes. The category of Loan is yes. The category of Contact is unknown. The value


In [ ]:
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig

#Загрузка модели Qwen 3.0 4B Instruct
model_name = "Qwen/Qwen3-4B-Instruct-2507"

device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {device}")

# Загрузка токенизаторов
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Загрузка модель с 16-битной квантизацией и автоматическим распределением по устройствам
base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    dtype=torch.bfloat16,
    device_map=device, # Explicitly force all model layers to the detected device (GPU if available)
    attn_implementation="sdpa"
)

if torch.cuda.is_available():
    print(f"Использовано памяти: {torch.cuda.memory_allocated() / 1e9:.2f} GB")

Using device: cuda


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


Loading weights:   0%|          | 0/398 [00:00<?, ?it/s]

Использовано памяти: 8.04 GB


In [ ]:
def create_prompt_missing(row, feature_names, target_name, prompt_config, tokenizer, missing_rate=0.0, seed=None):

    system_prompt = (
        f"You are a classifier. {prompt_config['task']}: "
        f"'{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'. "
        f"Answer with only one word: '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'."
    )

    input_text = row_to_text_template_with_missing(row, feature_names, target_name, missing_rate, seed)

    messages = [
        {"role": "system", "content": system_prompt},
        {"role": "user",
         "content": f"{prompt_config['entity']} information: {input_text}\n{prompt_config['question']}"},
    ]

    return tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True # добавление role assistant
    )

In [ ]:
import torch.nn.functional as F

def predict_single_with_prob(prompt, prompt_config, model, tokenizer, device, max_new_tokens=5):

    model_inputs = tokenizer([prompt], return_tensors="pt").to(device)

    with torch.no_grad():
        outputs = model.generate(
            model_inputs.input_ids,
            max_new_tokens=max_new_tokens,
            do_sample=False,
            pad_token_id=tokenizer.eos_token_id,
            output_scores=True,
            return_dict_in_generate=True
        )

    # Декодирование текста
    generated_ids = outputs.sequences[0][model_inputs.input_ids.shape[1]:]
    response = tokenizer.decode(generated_ids, skip_special_tokens=True).strip().lower()

    # Извлечение вероятностей для меток из prompt_config
    first_token_logits = outputs.scores[0][0]

    pos_id = tokenizer.encode(prompt_config['pos_label'], add_special_tokens=False)[0]
    neg_id = tokenizer.encode(prompt_config['neg_label'], add_special_tokens=False)[0]

    pos_logit = first_token_logits[pos_id]
    neg_logit = first_token_logits[neg_id]

    probs = F.softmax(torch.stack([pos_logit, neg_logit]), dim=0)
    prob_pos = probs[0].item()

    return response, prob_pos

# Тест
test_prompt = create_prompt_missing(test_df.iloc[0], feature_names, target_name, prompt_config, tokenizer)
response, prob = predict_single_with_prob(test_prompt, prompt_config, base_model, tokenizer, device)
print(f"Response: '{response}', Probability: {prob:.4f}")

The following generation flags are not valid and may be ignored: ['temperature', 'top_p', 'top_k']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


Response: 'yes', Probability: 0.9968


# 3. Fine-tuning с LoRA

## Подготовка данных для Fine-tuning

In [ ]:
def balance_dataset(train_df, target_name, seed=42):
    df_0 = train_df[train_df[target_name] == 0]
    df_1 = train_df[train_df[target_name] == 1]

    df_majority = df_0 if len(df_0) > len(df_1) else df_1
    df_minority = df_1 if len(df_0) > len(df_1) else df_0

    print(f"До балансировки:")
    print(f"{prompt_config['neg_label']}:  {len(df_majority)}")
    print(f"{prompt_config['pos_label']}: {len(df_minority)}")

    df_minority_up = resample(df_minority, replace=True,
                              n_samples=len(df_majority), random_state=seed)

    train_df_balanced = pd.concat([df_majority, df_minority_up])
    train_df_balanced = train_df_balanced.sample(frac=1, random_state=seed).reset_index(drop=True)

    print(f"\nПосле балансировки:")
    print(train_df_balanced[target_name].value_counts())

    return train_df_balanced

train_df_balanced = balance_dataset(train_df, target_name)

До балансировки:
no:  23953
yes: 3173

После балансировки:
y
1    23953
0    23953
Name: count, dtype: int64


In [ ]:
# Создание текстов для обучения

def create_training_dataset(df, feature_names, target_name, prompt_config, tokenizer, missing_rate=0.0):
    system_prompt = (
        f"You are a classifier. {prompt_config['task']}: "
        f"'{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'. "
        f"Answer with only one word: '{prompt_config['pos_label']}' or '{prompt_config['neg_label']}'."
    )
    texts = []

    for idx, (_, row) in enumerate(tqdm(df.iterrows(), total=len(df), desc=f"Dataset (missing={missing_rate:.0%})")):
        input_text = row_to_text_template_with_missing(row, feature_names, target_name, missing_rate, idx)
        target = prompt_config['pos_label'] if row[target_name] == 1 else prompt_config['neg_label']

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": f"{prompt_config['entity']} information: {input_text}\n{prompt_config['question']}"},
            {"role": "assistant", "content": target}
        ]

        text = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=False
        )
        texts.append(text)

    return texts


In [ ]:
def finetune_model(missing_rate, train_df_balanced, feature_names, target_name,
                   prompt_config, tokenizer, base_model, device,
                   num_epochs=3, batch_size=16):
    """
    Полный цикл fine-tuning для заданного missing_rate.
    Возвращает путь к лучшему checkpoint.
    """
    output_dir = f"{base_output_dir}_missing{int(missing_rate * 100):03d}"
    print(f"  Fine-tuning: missing_rate = {missing_rate:.0%}")
    print(f"  Output dir : {output_dir}")

    # Датасет
    train_texts = create_training_dataset(
        train_df_balanced, feature_names, target_name,
        prompt_config, tokenizer, missing_rate=missing_rate)
    train_dataset = Dataset.from_dict({"text": train_texts})

    def tokenize_fn(examples):
        return tokenizer(examples["text"], truncation=True, max_length=512, padding=False)

    tokenized_dataset = train_dataset.map(tokenize_fn, batched=True,
                                          remove_columns=train_dataset.column_names)

    # LoRA
    lora_config = LoraConfig(
        r=16, lora_alpha=32,
        target_modules=["q_proj", "k_proj", "v_proj", "o_proj",
                        "gate_proj", "up_proj", "down_proj"],
        lora_dropout=0.05, bias="none", task_type="CAUSAL_LM"
    )
    model_lora = get_peft_model(base_model, lora_config)
    model_lora.print_trainable_parameters()

    training_args = TrainingArguments(
        output_dir=output_dir,
        num_train_epochs=num_epochs,
        per_device_train_batch_size=batch_size,
        gradient_accumulation_steps=2,
        learning_rate=2e-4,
        bf16=True,
        logging_steps=50,
        save_strategy="epoch",
        optim="adamw_torch_fused",
        warmup_steps=50,
        max_grad_norm=1.0,
        weight_decay=0.01,
        report_to="none",
        dataloader_num_workers=4,
        group_by_length=True,
    )

    trainer = Trainer(
        model=model_lora,
        args=training_args,
        train_dataset=tokenized_dataset,
        data_collator=DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm=False)
    )

    t0 = time.time()
    trainer.train()
    print(f"Обучение завершено за {(time.time()-t0)/60:.1f} мин")

    trainer.save_model(output_dir)

    # Возвращаем базовую модель без LoRA весов (важно для следующего эксперимента)
    model_lora = model_lora.unload()
    torch.cuda.empty_cache()

    return output_dir

In [ ]:
import glob
import re
from peft import PeftModel

def evaluate_checkpoints_on_val(output_dir, val_df, feature_names, target_name,
                                 prompt_config, tokenizer, base_model, device,
                                 eval_missing_rate=0.0):

    checkpoints = sorted(
        glob.glob(f"{output_dir}/checkpoint-*"),
        key=lambda x: int(re.search(r'checkpoint-(\d+)', x).group(1))
    )
    print(f"Найдено {len(checkpoints)} checkpoints в {output_dir}")

    best_auc, best_checkpoint = 0.0, None
    for cp in checkpoints:
        model_eval = PeftModel.from_pretrained(base_model, cp).eval()

        y_true, y_prob = [], []
        for _, row in tqdm(val_df.iterrows(), total=len(val_df),
                           desc=f"Val eval {cp.split('/')[-1]}"):
            prompt = create_prompt_missing(
                row, feature_names, target_name, prompt_config, tokenizer,
                missing_rate=eval_missing_rate)
            _, prob = predict_single_with_prob(prompt, prompt_config, model_eval, tokenizer, device)
            y_true.append(row[target_name])
            y_prob.append(prob)

        y_pred = [1 if p > 0.5 else 0 for p in y_prob]
        auc = roc_auc_score(y_true, y_prob)
        print(f"  {cp.split('/')[-1]}  ROC-AUC={auc:.4f}")

        if auc > best_auc:
            best_auc = auc
            best_checkpoint = cp

        del model_eval
        torch.cuda.empty_cache()

    print(f"Лучший checkpoint: {best_checkpoint}  (ROC-AUC={best_auc:.4f})")
    return best_checkpoint

In [ ]:
def evaluate_on_test(best_checkpoint, test_df, feature_names, target_name,
                     prompt_config, tokenizer, base_model, device,
                     eval_missing_rate=0.0):
    model_finetuned = PeftModel.from_pretrained(base_model, best_checkpoint).eval()

    y_true, y_pred, y_prob = [], [], []
    t0 = time.time()

    for _, row in tqdm(test_df.iterrows(), total=len(test_df), desc="Test eval"):
        prompt = create_prompt_missing(
            row, feature_names, target_name, prompt_config, tokenizer,
            missing_rate=eval_missing_rate)
        response, prob = predict_single_with_prob(
            prompt, prompt_config, model_finetuned, tokenizer, device)
        y_true.append(row[target_name])
        y_pred.append(parse_prediction(response, prompt_config))
        y_prob.append(prob)

    elapsed = time.time() - t0
    metrics = compute_metrics(np.array(y_true), np.array(y_pred), np.array(y_prob))
    boot = bootstrap_metrics(np.array(y_true), np.array(y_pred), np.array(y_prob), n_iter=1000)

    print("Метрики:")
    for k, v in metrics.items():
        print(f"  {k}: {v:.4f}")
    print("Bootstrap (mean±std):")
    for k, v in boot.items():
        print(f"  {k}: {v}")

    del model_finetuned
    torch.cuda.empty_cache()

    return {
        "metrics": metrics,
        "bootstrap": boot,
        "time_total": elapsed,
        "time_per_sample": elapsed / len(y_true),
        "best_checkpoint": best_checkpoint,
        "eval_missing_rate": eval_missing_rate,
    }

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

all_results = {}

start_time = time.time()

for missing_rate in MISSING_RATES:
    tag = f"missing_{int(missing_rate * 100):03d}pct"
    print(f"Эксперимент: missing_rate = {missing_rate:.0%}")

    # 7.1 Fine-tuning
    output_dir = finetune_model(
        missing_rate=missing_rate,
        train_df_balanced=train_df_balanced,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        num_epochs=3,
        batch_size=16,
    )

    # 7.2 Выбор лучшего checkpoint по val
    best_cp = evaluate_checkpoints_on_val(
        output_dir=output_dir,
        val_df=val_df,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        eval_missing_rate=missing_rate,
    )

    # 7.3 Оценка на test
    result = evaluate_on_test(
        best_checkpoint=best_cp,
        test_df=test_df,
        feature_names=feature_names,
        target_name=target_name,
        prompt_config=prompt_config,
        tokenizer=tokenizer,
        base_model=base_model,
        device=device,
        eval_missing_rate=missing_rate,
    )

    all_results[tag] = result

print(f"Эксперимент завершен за {(time.time()-start_time)/60:.1f} мин")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Эксперимент: missing_rate = 0%
  Fine-tuning: missing_rate = 0%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing000


Dataset (missing=0%):   0%|          | 0/47906 [00:00<?, ?it/s]

Map:   0%|          | 0/47906 [00:00<?, ? examples/s]

trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
50,0.721683
100,0.192304
150,0.190850
200,0.187467
250,0.186579
300,0.187041
350,0.186985
400,0.184734
450,0.184216
500,0.183642


Обучение завершено за 89.8 мин
Найдено 3 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing000


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-1498:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-1498  ROC-AUC=0.9262


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-2996:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-2996  ROC-AUC=0.9130


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-4494:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-4494  ROC-AUC=0.9081
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing000/checkpoint-1498  (ROC-AUC=0.9262)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/9043 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.9195
  F1: 0.5132
  Accuracy: 0.7982
  Precision: 0.3575
  Recall: 0.9093
Bootstrap (mean±std):
  ROC-AUC: 0.9196±0.0038
  F1: 0.5135±0.0102
  Accuracy: 0.7982±0.0043
  Precision: 0.3578±0.0094
  Recall: 0.9095±0.0089
Эксперимент: missing_rate = 20%
  Fine-tuning: missing_rate = 20%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing020


Dataset (missing=20%):   0%|          | 0/47906 [00:00<?, ?it/s]

Map:   0%|          | 0/47906 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss


Step,Training Loss
50,0.848502
100,0.232840
150,0.229127
200,0.224327
250,0.223096
300,0.223342
350,0.220151
400,0.218731
450,0.218905
500,0.217593


Обучение завершено за 78.4 мин
Найдено 3 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing020


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-1498:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-1498  ROC-AUC=0.8960


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-2996:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-2996  ROC-AUC=0.9035


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-4494:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-4494  ROC-AUC=0.9038
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing020/checkpoint-4494  (ROC-AUC=0.9038)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/9043 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.8872
  F1: 0.5351
  Accuracy: 0.8416
  Precision: 0.4075
  Recall: 0.7788
Bootstrap (mean±std):
  ROC-AUC: 0.8873±0.0051
  F1: 0.5353±0.0112
  Accuracy: 0.8416±0.0039
  Precision: 0.4078±0.0112
  Recall: 0.7791±0.0129
Эксперимент: missing_rate = 50%
  Fine-tuning: missing_rate = 50%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing050


Dataset (missing=50%):   0%|          | 0/47906 [00:00<?, ?it/s]

Map:   0%|          | 0/47906 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
50,1.034594
100,0.251721
150,0.240152
200,0.237160
250,0.234127
300,0.231156
350,0.232476
400,0.230762
450,0.233728
500,0.230879


Обучение завершено за 61.4 мин
Найдено 3 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing050


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-1498:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-1498  ROC-AUC=0.8198


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-2996:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-2996  ROC-AUC=0.8300


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-4494:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-4494  ROC-AUC=0.8309
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing050/checkpoint-4494  (ROC-AUC=0.8309)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/9043 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.8336
  F1: 0.4395
  Accuracy: 0.7935
  Precision: 0.3220
  Recall: 0.6919
Bootstrap (mean±std):
  ROC-AUC: 0.8339±0.0064
  F1: 0.4400±0.0108
  Accuracy: 0.7935±0.0042
  Precision: 0.3225±0.0099
  Recall: 0.6925±0.0142
Эксперимент: missing_rate = 90%
  Fine-tuning: missing_rate = 90%
  Output dir : /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing090


Dataset (missing=90%):   0%|          | 0/47906 [00:00<?, ?it/s]

Map:   0%|          | 0/47906 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/peft/mapping_func.py:72: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


trainable params: 33,030,144 || all params: 4,055,498,240 || trainable%: 0.8145


Step,Training Loss
50,1.257148
100,0.183546
150,0.155996
200,0.166719
250,0.162635
300,0.160514
350,0.154735
400,0.155122
450,0.156794
500,0.149182


Обучение завершено за 59.4 мин
Найдено 3 checkpoints в /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing090


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-1498:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-1498  ROC-AUC=0.6572


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-2996:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-2996  ROC-AUC=0.6582


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Val eval checkpoint-4494:   0%|          | 0/9042 [00:00<?, ?it/s]

  checkpoint-4494  ROC-AUC=0.6761
Лучший checkpoint: /content/drive/MyDrive/finetuned_qwen_missing_Bank_missing090/checkpoint-4494  (ROC-AUC=0.6761)


/usr/local/lib/python3.12/dist-packages/peft/tuners/tuners_utils.py:285: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Test eval:   0%|          | 0/9043 [00:00<?, ?it/s]

Метрики:
  ROC-AUC: 0.6891
  F1: 0.2901
  Accuracy: 0.6877
  Precision: 0.1976
  Recall: 0.5454
Bootstrap (mean±std):
  ROC-AUC: 0.6893±0.0087
  F1: 0.2904±0.0094
  Accuracy: 0.6878±0.0048
  Precision: 0.1979±0.0074
  Recall: 0.5456±0.0154
Эксперимент завершен за 851.7 мин


In [ ]:
all_results

{'missing_000pct': {'metrics': {'ROC-AUC': np.float64(0.9195454497030704),
   'F1': 0.5132035209389171,
   'Accuracy': 0.7981864425522504,
   'Precision': 0.357487922705314,
   'Recall': 0.9092627599243857},
  'bootstrap': {'ROC-AUC': '0.9196±0.0038',
   'F1': '0.5135±0.0102',
   'Accuracy': '0.7982±0.0043',
   'Precision': '0.3578±0.0094',
   'Recall': '0.9095±0.0089'},
  'time_total': 2149.791994571686,
  'time_per_sample': 0.2377299562724412,
  'best_checkpoint': '/content/drive/MyDrive/finetuned_qwen_missing_Bank_missing000/checkpoint-1498',
  'eval_missing_rate': 0.0},
 'missing_020pct': {'metrics': {'ROC-AUC': np.float64(0.8871979952960003),
   'F1': 0.535064935064935,
   'Accuracy': 0.8416454716355192,
   'Precision': 0.4075173095944609,
   'Recall': 0.77882797731569},
  'bootstrap': {'ROC-AUC': '0.8873±0.0051',
   'F1': '0.5353±0.0112',
   'Accuracy': '0.8416±0.0039',
   'Precision': '0.4078±0.0112',
   'Recall': '0.7791±0.0129'},
  'time_total': 2099.447144508362,
  'time_per_

In [ ]:
print("\nBootstrap (mean±std):")
for tag, res in all_results.items():
    rate = tag.replace("missing_", "").replace("pct", "%").lstrip("0") or "0%"
    print(f"\n  missing={rate}")
    for k, v in res["bootstrap"].items():
        print(f"    {k}: {v}")


Bootstrap (mean±std):

  missing=%
    ROC-AUC: 0.9196±0.0038
    F1: 0.5135±0.0102
    Accuracy: 0.7982±0.0043
    Precision: 0.3578±0.0094
    Recall: 0.9095±0.0089

  missing=20%
    ROC-AUC: 0.8873±0.0051
    F1: 0.5353±0.0112
    Accuracy: 0.8416±0.0039
    Precision: 0.4078±0.0112
    Recall: 0.7791±0.0129

  missing=50%
    ROC-AUC: 0.8339±0.0064
    F1: 0.4400±0.0108
    Accuracy: 0.7935±0.0042
    Precision: 0.3225±0.0099
    Recall: 0.6925±0.0142

  missing=90%
    ROC-AUC: 0.6893±0.0087
    F1: 0.2904±0.0094
    Accuracy: 0.6878±0.0048
    Precision: 0.1979±0.0074
    Recall: 0.5456±0.0154


ROC-AUC (0.92): Наблюдается колоссальный рост производительности по сравнению с режимами zero-shot (0.54) и few-shot (0.59).
Дообучение позволило модели превратиться из случайного классификатора в высокоэффективный инструмент анализа банковского маркетинга. Даже при 20% пропусков модель удерживает показатель на уровне 0.89, что подтверждает её высокую устойчивость к частичной потере данных.

Recall (0.91): Исключительно высокий показатель полноты при полном наборе данных.
Модель успешно идентифицирует более 91% всех потенциальных клиентов, готовых подписаться на депозит. Это критически важно для маркетинговых кампаний, где важно не упустить целевого клиента. Однако при экстремальных 90% пропусков Recall падает до 0.55, что свидетельствует о потере ключевых индикаторов намерения клиента.

Precision (0.36): Точность остается умеренной, что указывает на склонность модели к «агрессивному» поиску.
Модель предпочитает пометить больше клиентов как потенциально заинтересованных, чтобы обеспечить максимальный охват, даже если это увеличивает количество ложных срабатываний. При 20% пропусков точность парадоксально возрастает до 0.41, что может быть связано с автоматической фильтрацией шума в данных.

Accuracy (0.80): Общая точность предсказаний достигла стабильного и высокого уровня.
Fine-tuning помог модели адаптироваться к дисбалансу классов в банковских данных, значительно превзойдя возможности базовой модели без обучения. Пик точности наблюдается при 20% пропусков (0.84), что указывает на оптимальный баланс между доступной информацией и обобщающей способностью модели.

Время эксперимента: ~ 14 ч 12 мин.

Использовано оперативная памяти графического процесса: 37.9/80GB.

Графический процессор A100 80GB.